### PR TITLE
TDL-19099: Add missing tap-tester tests

### PR DESCRIPTION
**Pull request by harshpatel4crest**
# Description of change
TDL-19099: Add missing tap-tester tests
- Added missing assertion in discovery test
- Added missing tap-tester tests
  - All fields test
  - Automatic fields test
  - Bookmark test

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/base.py [hunk 1]
```diff
@@ -361,8 +364,7 @@ def select_all_streams_and_fields(conn_id, catalogs, select_all_fields: bool = T
             non_selected_properties = []
             if not select_all_fields:
                 # get a list of all properties so that none are selected
-                non_selected_properties = schema.get('annotated-schema', {}).get(
-                    'properties', {}).keys()
+                non_selected_properties = set(schema.get('annotated-schema', {}).get('properties', {}).keys())
```
Comment (savan-chovatiya) [inline tests/base.py:367]: Is there any specific reason to change the above code by converting to a set?

File: tests/base.py [hunk 2]
```diff
@@ -380,10 +382,6 @@ def timedelta_formatted(self, dtime, days=0):
...
         return self.expected_metadata()[stream][self.REPLICATION_METHOD] == self.INCREMENTAL
 
-    def dt_to_ts(self, dtime):
-        for date_format in self.DATETIME_FMT:
-            try:
-                date_stripped = int(time.mktime(dt.strptime(dtime, date_format).timetuple()))
-                return date_stripped
-            except ValueError:
-                continue
+    def dt_to_ts(self, dtime, format):
```
Comment (savan-chovatiya) [inline tests/base.py:385]: Add function comment here.

File: tests/test_chargebee_all_fields.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+            'archived_at',
+            'attached_addons',
+            'avalara_sale_type',
+            'trial_end_action'
+        },
+        'subscriptions': {
+            'offline_payment_method',
+            'gift_id'
+        }
+    }
```
Comment (savan-chovatiya) [inline tests/test_chargebee_all_fields.py:226]: Please add comments with the reason for each field like why we have to skip it.  Add a comment at the map level if the reason is the same for all fields on the map.

File: tests/test_chargebee_all_fields.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        return 'chargebee_all_fields_test'
+
+    def all_fields_test_run(self):
+        """
+        • Verify no unexpected streams were replicated
+        • Verify that more than just the automatic fields are replicated for each stream.
+        • verify all fields for each stream are replicated
+        """
+
+        version = 'V1' if self.product_catalog_v1 else 'V2'
```
Comment (savan-chovatiya) [inline tests/test_chargebee_all_fields.py:238]: This variable looks unused. Can you please remove it?

File: tests/test_chargebee_all_fields.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+                # Verify that you get some records for each stream
+                self.assertGreater(record_count_by_stream.get(stream, -1), 0)
+
+                # verify all fields for a stream were replicated
+                self.assertGreater(len(expected_all_keys), len(expected_automatic_keys))
+                self.assertTrue(expected_automatic_keys.issubset(expected_all_keys), msg=f'{expected_automatic_keys-expected_all_keys} is not in "expected_all_keys"')
+
+                # get fields to remove for the version
+                stream_fields_as_per_version = self.fields_to_remove_V1.get(stream, set()) if self.product_catalog_v1 \
+                    else self.fields_to_remove_V2.get(stream, set())
```
Comment (savan-chovatiya) [inline tests/test_chargebee_all_fields.py:300]: Update the above line to the if-else condition for better readability.

File: tests/test_chargebee_all_fields.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+                # remove some fields as data cannot be generated / retrieved
+                fields = self.fields_to_remove_common.get(stream, set()) | stream_fields_as_per_version
+                for field in fields:
+                    expected_all_keys.remove(field)
```
Comment (savan-chovatiya) [inline tests/test_chargebee_all_fields.py:304]: Avoid the `for` loop here and directly use the subtraction(-)  operator as `expected_all_keys` and `fields` both look like a set. Also, update the variable name `fields` with a more self-explanatory word.

File: tests/test_chargebee_bookmark.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+from base import ChargebeeBaseTest
+import dateutil.parser
+
+class ChargebeeBookmarkTest(ChargebeeBaseTest):
+
+    def name(self):
+        return "chargebee_bookmark_test"
+
+    # subtract 12 hours from the bookmark in the state file
+    def get_simulated_states(self, state):
```
Comment (savan-chovatiya) [inline tests/test_chargebee_bookmark.py:13]: Add function comment here.

**Review by savan-chovatiya — CHANGES_REQUESTED** [review]

File: tests/test_chargebee_all_fields.py [hunk 8]
```diff
@@ -0,0 +1,316 @@
+from tap_tester import connections, runner, menagerie
+from base import ChargebeeBaseTest
+
+class ChargebeeAllFieldsTest(ChargebeeBaseTest):
+
+    # fields that are common between V1 and V2
```
Comment (dbshah1212) [inline tests/test_chargebee_all_fields.py:6]: A mismatch between field name and comment, do provide detailed comment if you are skipping these fields.

File: tests/test_chargebee_all_fields.py [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+
+    def all_fields_test_run(self):
+        """
+        • Verify no unexpected streams were replicated
+        • Verify that more than just the automatic fields are replicated for each stream.
+        • verify all fields for each stream are replicated
+        """
+
+        version = 'V1' if self.product_catalog_v1 else 'V2'
+        untestable_streams = {'quotes'} # For V2, we have 0 records for 'quotes' stream
```
Comment (dbshah1212) [inline tests/test_chargebee_all_fields.py:239]:
```suggestion
        untestable_streams_of_v2 = {'quotes'} # For V2, we have 0 records for 'quotes' stream
```

File: tests/test_chargebee_automatic_fields.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        return "chargebee_automatic_fields_test"
+
+    def automatic_fields_test_run(self):
+        """
+        Testing that all the automatic fields are replicated despite de-selecting them
+        - Verify that only the automatic fields are sent to the target.
+        - Verify that all replicated records have unique primary key values.
+        """
+
+        untestable_streams = {'quotes'} # For V2, we have 0 records for 'quotes' stream
```
Comment (dbshah1212) [inline tests/test_chargebee_automatic_fields.py:16]: if it is common across all tests then can we define it in the base.py?

File: tests/test_chargebee_all_fields.py (hunk 9 above)
Comment (harshpatel4crest) [reply tests/test_chargebee_all_fields.py]: Updated the variable name.

File: tests/test_chargebee_all_fields.py (hunk 8 above)
Comment (harshpatel4crest) [reply tests/test_chargebee_all_fields.py]: Updated the comment.

File: tests/test_chargebee_bookmark.py (hunk 7 above)
Comment (harshpatel4crest) [reply tests/test_chargebee_bookmark.py:12]: Added function comment.

File: tests/test_chargebee_all_fields.py (hunk 6 above)
Comment (harshpatel4crest) [reply tests/test_chargebee_all_fields.py]: Updated the code to directly subtract the fields to remove from expected fields.

File: tests/test_chargebee_all_fields.py (hunk 5 above)
Comment (harshpatel4crest) [reply tests/test_chargebee_all_fields.py]: Updated.

File: tests/test_chargebee_all_fields.py (hunk 4 above)
Comment (harshpatel4crest) [reply tests/test_chargebee_all_fields.py]: Right. Removed unused variable.

File: tests/test_chargebee_all_fields.py (hunk 3 above)
Comment (harshpatel4crest) [reply tests/test_chargebee_all_fields.py:228]: Added comment for removed fields during assertion.

File: tests/base.py (hunk 2 above)
Comment (harshpatel4crest) [reply tests/base.py:387]: Added comment for the function.

File: tests/base.py (hunk 1 above)
Comment (harshpatel4crest) [reply tests/base.py]: No, it was changed for `event_type` field, but now as we have updated the mechanism. Reverted the change.

File: tests/test_chargebee_all_fields.py [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+            'reference_authorization_id',
+            'reversal_transaction_id',
+            'validated_at',
+            'fraud_reason',
+            'amount_capturable',
+            'reference_transaction_id'
+        },
+    }
+    # fields to remove for V2
+    fields_to_remove_V2 = {
```
Comment (dbshah1212) [inline tests/test_chargebee_all_fields.py:100]: What is the reason behind skipping these fields? Do add comments for the same.

File: tests/test_chargebee_all_fields.py [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+        },
+        'promotional_credits': {
+            'reference'
+        },
+        'events': {
+            'user'
+        }
+    }
+    # fields to remove for V1
+    fields_to_remove_V1 = {
```
Comment (dbshah1212) [inline tests/test_chargebee_all_fields.py:177]: Provide a reason for skipping in the comment.

File: tests/test_chargebee_all_fields.py (hunk 11 above)
Comment (harshpatel4crest) [reply tests/test_chargebee_all_fields.py:100]: I have added general comments for all the streams [here](https://github.com/singer-io/tap-chargebee/pull/83/files#diff-1f08d51aa03a23c2c59e9bec09f6da13b0bbede459fe4a8a3a5979e05ea79892R6).

File: tests/test_chargebee_all_fields.py (hunk 12 above)
Comment (harshpatel4crest) [reply tests/test_chargebee_all_fields.py:177]: I have added general comments for all the streams [here](https://github.com/singer-io/tap-chargebee/pull/83/files#diff-1f08d51aa03a23c2c59e9bec09f6da13b0bbede459fe4a8a3a5979e05ea79892R6).

File: tests/test_chargebee_automatic_fields.py (hunk 10 above)
Comment (harshpatel4crest) [reply tests/test_chargebee_automatic_fields.py:16]: Previously, Kyle has suggested us not to add skipping stream in the base file as in future someone would not know the skipped stream in that test case. [Reference](https://github.com/singer-io/tap-snapchat-ads/pull/15#discussion_r864843867)

**Review by savan-chovatiya — APPROVED** [review]

**Review by dbshah1212 — APPROVED** [review]

**Review by KrisPersonal — APPROVED** [review]

**Review by RushT007 — CHANGES_REQUESTED** [review]

File: tests/base.py [hunk 13]
```diff
@@ -380,10 +383,7 @@ def timedelta_formatted(self, dtime, days=0):
     def is_incremental(self, stream):
         return self.expected_metadata()[stream][self.REPLICATION_METHOD] == self.INCREMENTAL
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/base.py]: Added docstrings for functions in base test file.

File: tests/test_chargebee_all_fields.py [hunk 14]
```diff
@@ -0,0 +1,318 @@
+from tap_tester import connections, runner, menagerie
+from base import ChargebeeBaseTest
+
+class ChargebeeAllFieldsTest(ChargebeeBaseTest):
+
+    # fields to remove that are common for V1 and V2
+    fields_to_remove_common = {
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/test_chargebee_all_fields.py:9]: Updated the comment for removal of fields in all fields test.

File: tests/test_chargebee_all_fields.py (hunk 12 above)
Comment (harshpatel4crest) [reply tests/test_chargebee_all_fields.py:162]: Updated the comment for removal of fields in all fields test.

File: tests/base.py (hunk 2 above)
Comment (harshpatel4crest) [reply tests/base.py:382]: Removed comment over the function and added docstring.

File: tap_chargebee/streams/base.py [hunk 15]
```diff
@@ -195,30 +195,38 @@ def sync_data(self):
                     break
 
             records = response.get('list')
-            
-            to_write = self.get_stream_data(records)
-            
+
+            # list of deleted records
```
Comment (savan-chovatiya) [inline tap_chargebee/streams/base.py:199]: Add a detailed comment like "List of deleted plans, addons, and coupons from the events endpoint"

File: tap_chargebee/streams/base.py [hunk 16]
```diff
@@ -195,30 +195,38 @@ def sync_data(self):
...
+            deleted_records = []
+
             if self.config.get('include_deleted') not in ['false','False', False]:
                 if self.ENTITY == 'event':
-                    for event in to_write:
+                    # parse "event_type" and deleted plan/addon/coupon from API records
```
Comment (savan-chovatiya) [inline tap_chargebee/streams/base.py:204]:
```suggestion
                    # Parse "event_type" from events records and collect deleted plan/addon/coupon from events
```

File: tap_chargebee/streams/base.py [hunk 17]
```diff
@@ -195,30 +195,38 @@ def sync_data(self):
...
+                    for record in records:
+                        event = record.get(self.ENTITY)
                         if event["event_type"] == 'plan_deleted':
                             Util.plans.append(event['content']['plan'])
                         elif event['event_type'] == 'addon_deleted':
                             Util.addons.append(event['content']['addon'])
                         elif event['event_type'] == 'coupon_deleted':
                             Util.coupons.append(event['content']['coupon'])
+                # we need additional transform for deleted records as "to_write" already contains transformed data
```
Comment (savan-chovatiya) [inline tap_chargebee/streams/base.py:213]:
```suggestion
                # We need additional transform for deleted records as "to_write" already contains transformed data
```

File: tap_chargebee/streams/base.py [hunk 18]
```diff
@@ -195,30 +195,38 @@ def sync_data(self):
...
                     for coupon in Util.coupons:
-                        to_write.append(coupon) 
+                        deleted_records.append(self.transform_record(coupon))
+
+            # get records from API response and transform
+            to_write = self.get_stream_data(records)
 
-            
             with singer.metrics.record_counter(endpoint=table) as ctr:
+                # combine transformed records and deleted records for "plan, addon and coupon"
```
Comment (savan-chovatiya) [inline tap_chargebee/streams/base.py:228]:
```suggestion
                # Combine transformed records and deleted data of  "plan, addon and coupon" collected from events endpoint
```

Comment (harshpatel4crest) [reply tap_chargebee/streams/base.py]: Updated comment.

File: tap_chargebee/streams/base.py (hunk 17 above)
Comment (harshpatel4crest) [reply tap_chargebee/streams/base.py]: Updated comment.

File: tap_chargebee/streams/base.py (hunk 16 above)
Comment (harshpatel4crest) [reply tap_chargebee/streams/base.py]: Updated comment.

File: tap_chargebee/streams/base.py (hunk 15 above)
Comment (harshpatel4crest) [reply tap_chargebee/streams/base.py]: Added comment as suggested.

**Review by kspeer825 — APPROVED** [review]
nice work

File: tests/test_chargebee_start_date.py [hunk 19]
```diff
@@ -18,13 +18,13 @@ def start_date_test_run(self):
         """Instantiate start date according to the desired data set and run the test"""
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/test_chargebee_start_date.py:18]: Skipped missing assertion streams at starting of the test case.

**Pull request merged**